android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/Worker.kt

```diff
@@ -33,4 +35,63 @@ interface Worker {
   /** Called when the worker is stopped.  */
   @JvmDefault
   fun onStop() {}
+
+  /**
+   * Allows overriding the threading type where the Worker.onStart/Worker.onStop will be called upon
+   */
+  @JvmDefault
+  fun threadingType(): ThreadingType = ThreadingType.CALLER_THREAD
+}
+
+/**
+ * Threading type where Worker.onStart/onStop will be running upon
+ */
+enum class ThreadingType {
+  BACKGROUND_COMPUTATION,
+  BACKGROUND_IO,
+  MAIN,
+  CALLER_THREAD
+}
+
+/**
+ * Worker.onStart/onStop duration information
+ */
+data class BinderDuration(
+  /**
+   * Worker class name
+   */
+  val workerName: String,
+
+  /**
+   * Worker event type (START/STOP)
+   */
+  val workerEvent: WorkerEvent,
+
+  /**
+   * Threading Type where Worker will operate
+   */
+  val threadingType: ThreadingType,
+
+  /**
+   * Thread name where Worker.onStart/onStop was called
+   */
+  val threadName: String,
+
+  /**
+   * Total binding duration in milli of Worker.onStart/onStop
+   */
+  val totalBindingDurationMilli: Long
+)
+
+/**
+ * Reports Worker total binding duration details
+ */
+interface RibWorkerBinderDurationListener {
+
+  /**
+   * Reports when Worker.onStart/Worker.onStop has completed
+   */
+  fun onBindCompleted(
```

Comment:
Still feels a bit random to me to have duration metrics on the API itself.

Also the fact that this would apply only to `Worker` is unconvicing. I mean, why are we reporting `Worker` duration on callback but not `Interactor` `didBecomeActive` duration on attach by router? It feels inconsistent. 

Have you given a thought about wrapping `Worker` instead? 

```kotlin
class MonitoredWorker(val worker: Worker) : Worker {
  fun onStartCompleted(duration: Long) {...}
  fun onStopCompleted(duration: Long) {...}
  
  override fun onStart(lifecycle: WorkerScopeProvider) {
    val duration = measureTimeMillis { worker.onStart(lifecycle) }
    onStartCompleted(time)
  }
  
  override fun onStop() {
    val duration = measureTimeMillis { worker.onStop() }
    onStopCompleted(time)
  }
}

Reply:
I agree on the lacking monitoring for Interactor's (will look into another PR for this).

The main reason to keep this coupled to Worker are:

- Would be great to have a complete view of duration of all Workers without changing current Worker usages. Wrapping is a possibility but that will require major refactor in order to get this Monitored Worker. 

-  Still this new monitoring is really tied to ThreadingType which I would like to pass on the callback as well with the thread name, worker class name (I know this will be obfuscated on release builds but still useful info)

Reply:
For this point
> Wrapping is a possibility but that will require major refactor in order to get this Monitored Worker.

Not really, since all would be needed was to call 

`worker.asMonitoredWorker()`

At binding site 

```
WorkerBinder.bind(worker.asMonitoredWorker())

fun Worker.asMonitoredWorker() = MonitoredWorker(this)
```

Reply:
For this point
> Wrapping is a possibility but that will require major refactor in order to get this Monitored Worker.

Not really, since all would be needed was to call 

`worker.asMonitoredWorker()`

At binding site 

```
WorkerBinder.bind(worker.asMonitoredWorker())

fun Worker.asMonitoredWorker() = MonitoredWorker(this)
```

Reply:
Still there could be cases where a Worker implementation extends other classes so this may not be suitable for all. 

Also the consumer of WorkerBinder will need to explicit opt in for this monitoring, so this will defer with the main goal of global a complete monitoring of all existing workers running